internal/internal_task_handlers.go

```diff
@@ -1291,22 +1291,31 @@ func getRetryBackoffWithNowTime(p *RetryPolicy, attempt int32, err error, now, e
 	if p.MaximumAttempts > 0 && attempt >= p.MaximumAttempts {
 		return noRetryBackoff // max attempt reached
 	}
+
+	var backoffInterval time.Duration
+	// Extract backoff interval from error if it is a retryable error.
+	// Not using errors.As() since we don't want to explore the whole error chain.
+	if applicationErr, ok := err.(*ApplicationError); ok {
+		backoffInterval = applicationErr.nextRetryDelay
+	}
+	// Calculate next backoff interval if the error did not contain the next backoff interval.
 	// attempt starts from 1
-	backoffInterval := time.Duration(float64(p.InitialInterval) * math.Pow(p.BackoffCoefficient, float64(attempt-1)))
-	if backoffInterval <= 0 {
-		// math.Pow() could overflow
-		if p.MaximumInterval > 0 {
+	if backoffInterval == time.Duration(0) {
+		backoffInterval = time.Duration(float64(p.InitialInterval) * math.Pow(p.BackoffCoefficient, float64(attempt-1)))
+		if backoffInterval <= 0 {
+			// math.Pow() could overflow
+			if p.MaximumInterval > 0 {
+				backoffInterval = p.MaximumInterval
+			} else {
+				return noRetryBackoff
+			}
+		}
```

Comment:
```suggestion
		if backoffInterval <= 0 {
			// math.Pow() could overflow
			if p.MaximumInterval > 0 {
				backoffInterval = p.MaximumInterval
			}
		}
```

And add after all this code (so it applies w/ next retry):

```go
if backoffInterval <= 0 {
	return noRetryBackoff
}
```

Reply:
hm, but do we want a negative `next retry` to disable retries?

Reply:
Well, non-positive, but that seems right in my head. What would happen today if this negative value was returned?

Reply:
hm good point,  the code will treat it as `noRetryBackoff` anyway if it is negative.